File: src/services/src/cpp/se_client.cpp
#include <ros/ros.h>
#include <string>

// Include service definitions
#include "services/AddTwoInts.h"

// Get two integers from the user
bool get_input(int& a, int& b) {
  std::string a_str, b_str;

  std::cout << "a: ";
  std::cin >> a_str;

  try {
    a = std::stoi(a_str);
  }
  catch (...) {
    return false;
  }

  std::cout << "b: ";
  std::cin >> b_str;

  try {
    b = std::stoi(b_str);
  }
  catch (...) {
    return false;
  }

  return true;
}

// Define node actions
int main(int argc, char* argv[]) {
  // Initialize the node
  ros::init(argc, argv, "se_client");

  // Create the client
  // - NodeHandle node
  // - Type of the service
  // - Name of the service
  ros::NodeHandle node;
  ros::ServiceClient client = node.serviceClient<services::AddTwoInts>("add_two_ints");

  while (ros::ok()) {
    // Create the request
    services::AddTwoInts srv;
    int a, b;
    if (!get_input(a, b)) {
      break;
    }

    srv.request.a = a;
    srv.request.b = b;

    // Send the request
    if (client.call(srv)) {
      std::cout << "sum: " << (long int)srv.response.sum << "\n";
    }
    else {
      ROS_ERROR("Failed to call the service");
    }
  }

  return 0;
}

File: src/services/src/py/se_client.py
#!/usr/bin/env python
import rospy
from rospy.exceptions import ROSInterruptException

# Import the service definitions
from services.srv import AddTwoInts, AddTwoIntsRequest, AddTwoIntsResponse

# Function to send reqest
def send_request(a, b):
	try:
		# Create a client
		# - Name of the service
		# - Type of the service
		client = rospy.ServiceProxy("add_two_ints", AddTwoInts)

		# Send a request
		response = client(a, b)

		return response.sum

	except rospy.ServiceException(e):
		print(f'Server error: {e}')

# Define node action
def node():
	# Initialize the node
	rospy.init_node("se_client", anonymous=False)

	# Wait for the server to start
	rospy.wait_for_service("add_two_ints")

	# Start sending requests
	while not rospy.is_shutdown():
		try:
			a = int(input("a: "))
			b = int(input("b: "))
		except:
			break

		sum = send_request(a, b)
		print(f'sum: {sum}')


# Start the node if the file is executed
if __name__ == "__main__":
	try:
		node()
	except ROSInterruptException:
		pass
